### PR TITLE
[REVIEW] Fix issue when different dtype values are passed to `.cat.add_categories`

### PR DESCRIPTION
**Pull request by galipremsagar**
Fixes: #5264

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf/cudf/core/column/string.py [hunk 1]
```diff
@@ -2182,6 +2182,19 @@ def deserialize(cls, header, frames):
         )
         return col
 
+    def can_cast_safely(self, to_dtype):
+        # import pdb;pdb.set_trace()
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py:2186]: Remove dead code

```suggestion
```

File: python/cudf/cudf/core/column/categorical.py [hunk 2]
```diff
@@ -56,9 +56,28 @@ def as_unordered(self, **kwargs):
...
+        common_dtype = np.find_common_type(
+            [old_categories.dtype, new_categories.dtype], []
+        )
+
+        if new_categories.dtype != common_dtype:
+            new_categories = new_categories.astype(common_dtype)
+        elif old_categories.dtype != common_dtype:
+            old_categories = old_categories.astype(common_dtype)
+
+        already_included = set(new_categories) & set(old_categories)
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/categorical.py:70]: We don't want to convert the GPU columns to host `set` objects here.

File: python/cudf/cudf/core/column/categorical.py [hunk 3]
```diff
@@ -56,9 +56,28 @@ def as_unordered(self, **kwargs):
 
     def add_categories(self, new_categories, **kwargs):
         new_categories = column.as_column(new_categories)
-        new_categories = self._column.categories.append(new_categories)
+        old_categories = self._column.categories
+
+        common_dtype = np.find_common_type(
+            [old_categories.dtype, new_categories.dtype], []
+        )
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/categorical.py:63]: Are we sure we want this behavior? I.E. if someone has a dictionary of `['a', 'b', 'c']` and calls `add_categories([1, 2, 3])` the result in Pandas would be `['a', 'b', 'c', 1, 2, 3]` but our result would be `['a', 'b', 'c', '1', '2', '3']` which is somewhat unexpected behavior.

Should we consider throwing on mixed dtype instead? That way if a user is okay with the behavior they can explicitly convert.

File: python/cudf/cudf/core/column/categorical.py [hunk 4]
```diff
@@ -167,7 +186,7 @@ def _set_categories(self, new_categories, **kwargs):
 
         from cudf import DataFrame, Series
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/categorical.py:187]: Move this import out of the function.

File: python/cudf/cudf/core/column/categorical.py (hunk 3 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/categorical.py:74]: Changed this behavior to raise an error for mixed types.

File: python/cudf/cudf/core/column/categorical.py [hunk 5]
```diff
@@ -56,18 +56,42 @@ def as_unordered(self, **kwargs):
...
+
+        if (
+            old_categories.dtype == "object"
+            and new_categories.dtype != "object"
+        ) or (
+            old_categories.dtype != "object"
+            and new_categories.dtype == "object"
+        ):
+            raise TypeError(
+                "cudf doesnot support mixed types, please type-cast \
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/categorical.py:69]:
```suggestion
                "cudf does not support mixed types, please type-cast \
```

File: python/cudf/cudf/core/column/categorical.py [hunk 6]
```diff
@@ -56,18 +56,42 @@ def as_unordered(self, **kwargs):
...
         new_categories = column.as_column(new_categories)
-        new_categories = self._column.categories.append(new_categories)
+        old_categories = self._column.categories
+
+        if (
+            old_categories.dtype == "object"
+            and new_categories.dtype != "object"
+        ) or (
+            old_categories.dtype != "object"
+            and new_categories.dtype == "object"
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/categorical.py:66]: Better opportunity to short circuit in the case of all numeric:
```suggestion
           new_categories.dtype == "object"
            and old_categories.dtype != "object"
```

File: python/cudf/cudf/core/column/categorical.py [hunk 7]
```diff
@@ -281,9 +281,35 @@ def add_categories(self, new_categories, **kwargs):
...
+                    new_categories to the same type as existing categories."
+            )
+        common_dtype = np.find_common_type(
+            [old_categories.dtype, new_categories.dtype], []
+        )
+
+        if new_categories.dtype != common_dtype:
+            new_categories = new_categories.astype(common_dtype)
+        elif old_categories.dtype != common_dtype:
+            old_categories = old_categories.astype(common_dtype)
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/categorical.py:304]: Can just use zero-copy typecasting here:
```suggestion
        new_categories = new_categories.astype(common_dtype, copy=False)
        old_categories = old_categories.astype(common_dtype, copy=False)
```

File: python/cudf/cudf/core/column/categorical.py [hunk 8]
```diff
@@ -281,9 +281,33 @@ def add_categories(self, new_categories, **kwargs):
...
+        new_categories = new_categories.astype(common_dtype, copy=False)
+        old_categories = old_categories.astype(common_dtype, copy=False)
+
+        if old_categories.isin(new_categories).any():
+            raise ValueError("new categories must not include old categories")
+
+        new_categories = old_categories.append(new_categories)
         out_col = self._column
         if not self._categories_equal(new_categories, **kwargs):
+            kwargs["current_cats"] = old_categories
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/categorical.py:310]: kwargs are getting a bit overloaded here. Could we make a shallow copy of `self` and then change the categories on that and call `_set_categories` on that?

File: CHANGELOG.md [hunk 9]
```diff
@@ -10,6 +10,8 @@
 
 ## Bug Fixes
 
+- PR #5271 Fix issue when different dtype values are passed to `.cat.add_categories`
+
```
Comment (galipremsagar) [inline CHANGELOG.md:14]:
```suggestion
```

File: python/cudf/cudf/core/column/categorical.py (hunk 8 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/categorical.py]: Updated the `_set_categories` to actually take in current_categories as a parameter

**Pull request merged**